vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala

```diff
@@ -43,7 +43,12 @@ import geotrellis.util.GetComponent
   * @groupprio reproject 2
   */
 trait RasterSource extends CellGrid with AutoCloseable with Serializable {
+  // sfitch: As with `targetCellType` below, this feels like API leakage, and seems to be captured already by
+  // the `XXXResampleRasterSource` classes.
   def resampleMethod: Option[ResampleMethod]
+
+  // sfitch: what if we have `RasterSource`s that aren't identified by a URI? e.g. `InMemoryRasterSource`?
+  // Also: should we consider more strongly typing this?
```

Comment:
:+1: I think we're in for something like:

```scala 
 trait NamedRasterSource { self: RasterSource => 
   def uri: ???
}
```

`MosaicRasterSource` is another thing that doesn't have a URI sadly.

Reply:
@pomadchin @echeipesh Can one of you remind me of which cases won't allow us to use `java.net.URI`?

Reply:
S3 URIs allow characters that are not allowed by `java.net.URI`. We ran into this problem when using pre-signed URIs in Raster Foundry as well as having `|` be part of the S3 prefix for some reason :(

Reply:
Is this going too far?:

```scala
trait IdentifiedRasterSource { self: RasterSource =>
  def id: RasterSourceIdentifier
}

trait RasterSourceIdentifier {
  def connect: ???
  def asString: String
}

case class PathIdientifier(path: String) extends RasterSourceIdentifier
case class URIIdientifier(uri: URI) extends RasterSourceIdentifier
import com.amazonaws.services.s3.AmazonS3URI
case class S3URIIdentified(s3URI: AmazonS3URI) extends RasterSourceIdentifier

```

If so, I'd suggest this instead:

```scala
trait IdentifiedRasterSource { self: RasterSource =>
  def id: RasterSourceIdentifier
}

// Tagged type pattern...
trait ValidRSID
type RasterSourceIdentifier = String with ValidRSID
object RasterSourceIdentifier {
  def apply(path: String): RasterSourceIdentifier = {
    if (/* path is valid URI or S3 URI */) 
      path.asInstanceOf[RasterSourceIdentifier]
    else
      // Do something obnoxious....
  }
}
```